File: 2016/day1.js
let directions = require('./input/day1')

let heading = 0
let n = 0
let e = 0
let locations = [[0, 0]]

function changeHeading (turn) {
  if (turn === 'R') {
    heading += 90
  } else {
    heading -= 90
  }
  heading += 360
  heading %= 360
}
function sharedVertical () {
  return locations.filter(location => location[1] === e)
}
function sharedHorizontal () {
  return locations.filter(location => location[0] === n)
}
function check(possibleMatches, x, y) {
  if (possibleMatches.some(match => match[0] === x && match[1] === y)) {
    console.log("DONE", x, y)
  } else {
    locations.push([x, y])
  }
}
function goNorth (distance, location) {
  n += distance
  let possibleMatches = sharedVertical()
  for (var i = location[0]; i < n; i++) {
    check(possibleMatches, i, e)
  }
}
function goEast (distance, location) {
  e += distance
  let possibleMatches = sharedHorizontal()
  for (var i = location[1]; i < e; i++) {
    check(possibleMatches, n, i)
  }
}
function goSouth (distance, location) {
  n -= distance
  let possibleMatches = sharedVertical()
  for (var i = location[0]; i > n; i--) {
    check(possibleMatches, i, e)
  }
}
function goWest (distance, location) {
  e -= distance
  let possibleMatches = sharedHorizontal()
  for (var i = location[1]; i > e; i--) {
    check(possibleMatches, n, i)
  }
}
function advance (distance) {
  let location = locations[locations.length - 1]
  if (heading === 0) {
    goNorth(distance, location)
  } else if (heading === 90) {
    goEast(distance, location)
  } else if (heading === 180) {
    goSouth(distance, location)
  } else if (heading === 270) {
    goWest(distance, location)
  }
}
function step (direction) {
  let turn = direction.slice(0,1)
  let distance = parseInt(direction.slice(1))
  changeHeading(turn)
  advance(distance)
  console.log('at ', n, e)
}
directions.some(step)
console.log(Math.abs(n) + Math.abs(e))

File: 2016/day1.rb
directions = File.read('./input/day1.txt').strip.split(', ')

def nextLocation(heading, distance, location)
  case heading
  when 0
    { y: location[:y] + distance, x: location[:x] }
  when 90
    { y: location[:y], x: location[:x] + distance }
  when 180
    { y: location[:y] - distance, x: location[:x] }
  when 270
    { y: location[:y], x: location[:x] - distance }
  end
end

def parseMove(move, heading)
  direction, distance = move.split('', 2)
  heading = ((direction === 'R' ? heading + 90 : heading - 90) + 360) % 360
  [distance.to_i, heading]
end

heading = 0
destination = directions.reduce({ x: 0, y: 0 }) do |location, move|
  distance, heading = parseMove(move, heading)
  nextLocation(heading, distance, location)
end

puts destination[:y].abs + destination[:x].abs

def step(heading, location)
  case heading
  when 0
    { y: location[:y] + 1, x: location[:x] }
  when 90
    { y: location[:y], x: location[:x] + 1 }
  when 180
    { y: location[:y] - 1, x: location[:x] }
  when 270
    { y: location[:y], x: location[:x] - 1 }
  end
end

heading = 0
visited = {}
destination = directions.reduce({ x: 0, y: 0 }) do |location, move|
  done = false
  distance, heading = parseMove(move, heading)
  distance.times do
    location = step(heading, location)
    if visited[location]
      done = true
      break(location)
    else
      visited[location] = true
    end
  end
  if done 
    break(location) 
  else 
    location 
  end
end
puts destination[:y].abs + destination[:x].abs
